src/main/java/org/folio/service/finance/transaction/EncumbranceService.java

```diff
@@ -88,27 +90,34 @@ public Future<Void> createOrUpdateEncumbrances(EncumbrancesProcessingHolder hold
   }
 
   public Future<Void> createEncumbrances(List<EncumbranceRelationsHolder> relationsHolders, RequestContext requestContext) {
-    List<Future<Void>> futureList = new ArrayList<>();
-    Future<Void> future = Future.succeededFuture();
-    for (EncumbranceRelationsHolder holder : relationsHolders) {
-      // TODO: fix sequential processing with parallel using semaphores
-      future = future.compose(v -> transactionService.createTransaction(holder.getNewEncumbrance(), requestContext)
-        .map(transaction -> {
-          holder.getFundDistribution().setEncumbrance(transaction.getId());
-          return null;
-        })
-        .recover(fail -> {
-          checkCustomTransactionError(fail);
-          throw new CompletionException(fail);
-        })
-        .mapEmpty()
-      );
-      futureList.add(future);
-    }
-    return GenericCompositeFuture.join(futureList)
-      .mapEmpty();
+    Semaphore semaphore = new Semaphore(SEMAPHORE_MAX_ACTIVE_THREADS, requestContext.getContext().owner());
+
+    return requestContext.getContext().owner()
+      .<List<Future<Transaction>>>executeBlocking(event -> {
+
+        List<Future<Transaction>> futures = new ArrayList<>();
+
+        for (EncumbranceRelationsHolder holder : relationsHolders) {
+          Future<Transaction> future = transactionService.createTransaction(holder.getNewEncumbrance(), requestContext);
+          futures.add(future);
+          semaphore.acquire(() -> future.map(transaction -> {
```

Comment:
Can we refactor to this way to make semaphore declaration consistent across all places:
semaphore.acquire(() -> future
  .onComplete(asyncResult -> semaphore.release()));

And future map logic move to the same future declaration here transactionService.createTransaction(holder.getNewEncumbrance(), requestContext)

Or semaphores will not work after this refactoring?

Reply:
refactored